src/main/java/org/openstreetmap/atlas/checks/validation/linear/lines/WaterWayCheck.java

```diff
@@ -0,0 +1,472 @@
+import org.openstreetmap.atlas.geography.atlas.items.Edge;
+import org.openstreetmap.atlas.geography.atlas.items.Line;
+import org.openstreetmap.atlas.geography.atlas.items.LineItem;
+import org.openstreetmap.atlas.geography.atlas.items.LocationItem;
+import org.openstreetmap.atlas.geography.atlas.items.Node;
+import org.openstreetmap.atlas.geography.atlas.items.Point;
+import org.openstreetmap.atlas.tags.LayerTag;
+import org.openstreetmap.atlas.tags.SyntheticBoundaryNodeTag;
+import org.openstreetmap.atlas.tags.filters.TaggableFilter;
+import org.openstreetmap.atlas.utilities.collections.Iterables;
+import org.openstreetmap.atlas.utilities.collections.Sets;
+import org.openstreetmap.atlas.utilities.configuration.Configuration;
+import org.openstreetmap.atlas.utilities.scalars.Angle;
+import org.openstreetmap.atlas.utilities.scalars.Distance;
+
+/**
+ * This check flags waterways that do not have a sink (i.e., are not connected to another waterway),
+ * are circular (so first node and last node are the same), or cross another waterway with the same
+ * layer. It also looks for ways that may be going uphill (requires elevation data, see
+ * {@link ElevationUtilities}
+ *
+ * @author Taylor Smock
+ */
+public class WaterWayCheck extends BaseCheck<Long>
+{
+    /**
+     * This comparator takes a polyline and segments, and compares the segments based off of the
+     * location of the segment in the line.
+     *
+     * @author Taylor Smock
+     */
+    private static class SegmentIndexComparator implements Comparator<Segment>
+    {
+        private final List<Segment> lineSegments;
+
+        /**
+         * Initialize with a specific polyline to check
+         *
+         * @param line
+         *            The line to check the index of segments against
+         */
+        SegmentIndexComparator(final PolyLine line)
+        {
+            this.lineSegments = line.segments();
+        }
+
+        @Override
+        public int compare(final Segment segment1, final Segment segment2)
+        {
+            final Segment segment1real = this.lineSegments.stream().filter(segment1::equals)
+                    .findFirst().orElse(null);
+            final Segment segment2real = this.lineSegments.stream().filter(segment2::equals)
+                    .findFirst().orElse(null);
+            final int segment1index = this.lineSegments.indexOf(segment1real);
+            final int segment2index = this.lineSegments.indexOf(segment2real);
+            return segment1index - segment2index;
+        }
+    }
+
+    private static final long serialVersionUID = 2877101774578564205L;
+    private static final String WATERWAY_SINK_TAG_FILTER_DEFAULT = "natural->sinkhole|waterway->tidal_channel,drain|manhole->drain";
+    private static final String WATERWAY_TAG_FILTER_DEFAULT = "waterway->river,stream,tidal_channel,canal,drain,ditch,pressurised";
+
+    private static final String DEFAULT_VALID_OCEAN_TAGS = "natural->strait,channel,fjord,sound,bay|"
```

Comment:
Is this copied from OceanBleedingCheck? https://github.com/osmlab/atlas-checks/blob/dev/src/main/java/org/openstreetmap/atlas/checks/validation/intersections/OceanBleedingCheck.java#L40

If so, lets pull them both out into a tag util that can be shared

Reply:
IIRC, it is. I think I copied it, with the intention of seeing if I would have to change it. I never did.

Reply:
I created another Utility class specifically for common filters.